File: terraform/scripts/loadtest.js
import http from 'k6/http';
import { check, sleep, group } from 'k6';
import { Rate, Trend, Counter } from 'k6/metrics';
import { SharedArray } from 'k6/data';
import { textSummary } from 'https://jslib.k6.io/k6-summary/0.0.1/index.js';

const BASE_URL = __ENV.BASE_URL || 'http://localhost';
const TEST_TYPE = __ENV.TEST_TYPE || 'load';

const TIMEZONES = new SharedArray('timezones', function() {
  return [
    'America/New_York', 'Europe/London', 'Asia/Tokyo', 'Australia/Sydney',
    'Asia/Dubai', 'Asia/Singapore', 'America/Sao_Paulo', 'Asia/Kolkata',
    'Europe/Paris', 'America/Los_Angeles', 'Asia/Hong_Kong', 'Europe/Berlin'
  ];
});

// SLO-friendly metrics to populate the Grafana SLO Compliance dashboard
const sliErrorBudgetBurn = new Rate('sli_error_budget_burn');
const sliLatencyVsSloRatio = new Trend('sli_latency_vs_slo_ratio');
const sliRequestCount = new Counter('sli_request_count');

function recordSli(res, endpoint, latencySloMs) {
  const latency = res.timings.duration;
  const success = res.status >= 200 && res.status < 400;
  const tags = { endpoint };

  sliRequestCount.add(1, tags);
  sliErrorBudgetBurn.add(!success, tags);

  if (latencySloMs > 0) {
    sliLatencyVsSloRatio.add(latency / latencySloMs, tags);
  }
}

const profiles = {
  // Expected Normal Load
  load: {
    stages: [
      { duration: '2m', target: 500 }, // Start with 500 users
      { duration: '10m', target: 500 }, // Stay at 500 users
      { duration: '2m', target: 0 }, // Ramp down to 0 users
    ],
    thresholds: {
      'http_req_duration': ['p(95)<500', 'p(99)<1000'],  // SLO targets
      'http_req_failed': ['rate<0.001'],  // Error rate < 0.1%
      'sli_error_budget_burn': ['rate<0.001'],
      'sli_latency_vs_slo_ratio': ['p(95)<1.0'],
    },
  },

  // Stress Testing to find breaking points
  stress: {
    stages: [
      { duration: '2m', target: 500 }, // Start with 500 users
      { duration: '5m', target: 750}, // Ramp up to 750 users
      { duration: '5m', target: 1250}, // Ramp up to 1250 users
      { duration: '5m', target: 2500}, // Ramp up to 2500 users
      { duration: '5m', target: 5000}, // Ramp up to 5000 users
      { duration: '10m', target: 5000}, // Stay at 5000 users
      { duration: '5m', target: 0 }, // Recover
    ],
    thresholds: {
      'http_req_duration': ['p(95)<2000', 'p(99)<5000'],  // Relaxed SLO targets
      'http_req_failed': ['rate<0.05'],  // Error rate < 5%
    },
  },

  // Spike Testing for sudden traffic bursts
  spike: {
    stages: [
      { duration: '1m', target: 500 }, // Baseline
      { duration: '5m', target: 5000 }, // Spike to 5000 users
      { duration: '10m', target: 5000 }, // Stay at spike
      { duration: '2m', target: 500 }, // Drop back to baseline
      { duration: '1m', target: 500 }, // Stay at baseline
      { duration: '1m', target: 0 }, // Recover
    ],
    thresholds: {
      'http_req_duration': ['p(95)<2000', 'p(99)<5000'],
      'http_req_failed': ['rate<0.10'],  // Error rate < 10%
    },
  },

  // Soak Testing for long-term stability
  soak: {
    stages: [
      { duration: '5m', target: 500 }, // Start with 500 users
      { duration: '3h', target: 500 }, // Hold for 3 hours
      { duration: '5m', target: 0 }, // Ramp down to 0 users
    ],
    thresholds: {
      'http_req_duration': ['p(95)<500', 'p(99)<1000'],  // Standard SLO targets
      'http_req_failed': ['rate<0.01'],  // Error rate < 1%
      'sli_error_budget_burn': ['rate<0.001'],
      'sli_latency_vs_slo_ratio': ['p(95)<1.0'],
    },
  },
};

const selectedProfile = profiles[TEST_TYPE];

export let options = {
  ...selectedProfile,
  tags: {
    test_type: TEST_TYPE,
  },
};

export default function () {
  let rand = Math.random();

  if (rand < 0.8) {
    // World Clocks API Endpoint. 80% of traffic, expected to be the most common endpoint hit by users
    group('world clocks endpoint', () => {
      let res = http.get(`${BASE_URL}/api/world-clocks`, {
        tags: { name: 'world-clocks', endpoint: 'world-clocks' },
      });
      check(res, {
        'world clocks status 200': (r) => r.status === 200,
        'world clocks latency < 500ms': (r) => r.timings.duration < 500,
      });
      recordSli(res, 'world-clocks', 500);
    });
  } else if (rand < 0.9) {
    // Homepage Endpoint. 10% of traffic, expected to be common but less than API calls
    group('homepage endpoint', () => {
      let res = http.get(`${BASE_URL}/`, {
        tags: { name: 'homepage', endpoint: 'homepage' },
      });
      check(res, {
        'homepage status 200': (r) => r.status === 200,
        'homepage latency < 500ms': (r) => r.timings.duration < 500,
      });
      recordSli(res, 'homepage', 500);
    });
  } else {
    // Current Time API Endpoint. Also 10% of traffic, expected to be common but less than world clocks endpoint
    group('time endpoint', () => {
      let tz = TIMEZONES[Math.floor(rand * TIMEZONES.length)];
      let res = http.get(`${BASE_URL}/api/time?timezone=${tz}`, {
        tags: { name: 'time', endpoint: 'time' },
      });
      check(res, {
        'time status 200': (r) => r.status === 200,
        'time latency < 500ms': (r) => r.timings.duration < 500,
      });
      recordSli(res, 'time', 500);
    });
  }

  sleep(1 + rand * 2); // Think time 1-3s
}

export function handleSummary(data) {
  return {
    '/tmp/summary.json': JSON.stringify(data),
    'stdout': textSummary(data, { indent: ' ', enableColors: true }),
  };
}

File: terraform/scripts/loadtest.py
#!/usr/bin/env python3
"""
Load test script for World Clock API
Generates realistic traffic patterns to populate DB metrics
"""
import requests
import random
import time
from concurrent.futures import ThreadPoolExecutor, as_completed

# Update this to your actual endpoint
BASE_URL = "https://your.domain/api"  # or http://localhost:5000 for local

TIMEZONES = [
    "America/New_York", "Europe/London", "Asia/Tokyo", "Australia/Sydney",
    "Asia/Dubai", "Asia/Singapore", "America/Sao_Paulo", "Asia/Kolkata",
    "Europe/Paris", "America/Los_Angeles", "Asia/Hong_Kong", "Europe/Berlin",
    "Invalid/Zone",  # Intentional error to generate 400s
]

ENDPOINTS = [
    ("/world-clocks", {}),
    ("/timezones", {}),
    ("/time", {"timezone": lambda: random.choice(TIMEZONES)}),
]

def make_request(session, run_num):
    """Single request with random endpoint/timezone"""
    endpoint, params = random.choice(ENDPOINTS)
    
    # Build params if it's a callable (for random values)
    request_params = {k: v() if callable(v) else v for k, v in params.items()}
    
    try:
        start = time.time()
        resp = session.get(f"{BASE_URL}{endpoint}", params=request_params, timeout=5)
        latency = (time.time() - start) * 1000
        
        status_emoji = "✅" if resp.status_code < 400 else "❌"
        print(f"{status_emoji} [{run_num:4d}] {endpoint:20s} → {resp.status_code} ({latency:.1f}ms)")
        
        return {
            "status": resp.status_code,
            "latency": latency,
            "endpoint": endpoint
        }
    except Exception as e:
        print(f"❌ [{run_num:4d}] Request failed: {e}")
        return {"status": 0, "latency": 0, "endpoint": endpoint}

def run_load_test(duration_seconds=300, concurrent_users=10):
    """
    Run load test
    - duration_seconds: how long to run (default 5min)
    - concurrent_users: parallel requests
    """
    print(f"🚀 Starting load test: {duration_seconds}s, {concurrent_users} concurrent users")
    print(f"   Target: {BASE_URL}")
    print("=" * 70)
    
    session = requests.Session()
    start_time = time.time()
    request_count = 0
    
    with ThreadPoolExecutor(max_workers=concurrent_users) as executor:
        futures = []
        
        while (time.time() - start_time) < duration_seconds:
            # Submit a new request
            future = executor.submit(make_request, session, request_count)
            futures.append(future)
            request_count += 1
            
            # Small delay between submissions (adjust for desired RPS)
            time.sleep(random.uniform(0.05, 0.2))  # ~5-20 RPS per user
            
            # Collect completed futures periodically
            if len(futures) > 100:
                done_futures = [f for f in futures if f.done()]
                futures = [f for f in futures if not f.done()]
        
        # Wait for remaining
        print("\n⏳ Waiting for remaining requests...")
        for future in as_completed(futures):
            future.result()
    
    elapsed = time.time() - start_time
    print("=" * 70)
    print(f"✅ Load test complete!")
    print(f"   Total requests: {request_count}")
    print(f"   Duration: {elapsed:.1f}s")
    print(f"   Avg RPS: {request_count/elapsed:.1f}")

if __name__ == "__main__":
    # Quick test (30 seconds, 5 users)
    run_load_test(duration_seconds=30, concurrent_users=5)
    
    # For serious load, use:
    # run_load_test(duration_seconds=600, concurrent_users=20)